docs/build-and-publish-pipeline.md

```diff
@@ -0,0 +1,68 @@
+## Application Services Build and Publish Pipeline
+
+This document provides an overview of the build-and-publish pipeline used to make our work
+in this repo available to consuming applications. It's intended both to document the pipeline
+for development and maintenance purposes, and to serve as a basic analysis of the integrity
+protections that it offers (so you'll notice there are notes and open questions in place where
+we haven't fully hashed out all those details).
+
+The key points:
+
+* We use [Cargo](https://github.com/rust-lang/cargo) for building and testing the core Rust code in isolation,
+  [Gradle](https://gradle.org/) with [rust-android-gradle](https://github.com/mozilla/rust-android-gradle)
+  for combining Rust and Kotlin code into Android components and running tests against them,
+  and [Carthage](https://github.com/Carthage/Carthage) for combining Rust and Swift code into iOS components.
+* [TaskCluster](../automation/taskcluster/README.md) runs on every pull-request, tag
+  and push to master, to ensure Android artifacts build correctly and to execute their
+  tests via gradle.
+* [CircleCI](../.circleci/config.yml) runs on every branch, pull-request, and tag,
+  to execute lint checks and automated tests at the rust level.
+    * TODO: how do we run automated tests for the Swift wrapper code?
+* Releases are made by [manually creating a new tag](./howtos/cut-a-new-release.md),
+  which triggers various CI jobs:
+    * [CircleCI](../.circleci/config.yml) is used to build an iOS binary release on every tag,
+      and publish it as a GitHub release artifcact.
+    * [TaskCluster](../automation/taskcluster/README.md) is used to:
+        * Build an Android binary release.
+        * Publish it to nalexander@'s personal bintray (although this will soon change to publish
+           to https://maven.mozilla.org).
+        * Upload symbols to [Socorro](https://wiki.mozilla.org/Socorro).
+           * TODO: does this mean the symbols are only valid for debugging Android crashes, not on iOS?
+
+For Android consumers these are the steps by which Application Services code becomes available,
+and the integrity-protection mechanisms that apply at each step:
+
+1. Code is developed in branches and lands on `master` via pull request.
+    * GitHub branch protection prevents code being pushed to `master` without review.
+      * TODO: we should consider requiring signed tags before merge to `master`.
+    * CircleCI and TaskCluster run automated tests against the code, but do not have
```

Comment:
Something we should talk about is how we disabled running TC tasks for the general public (only collaborators can trigger TC tasks with their PRs).